docs/architecture/key-management-services.md

```diff
@@ -4,79 +4,75 @@
 
 - [edit live](https://mermaid.live/edit)
 
-This document describes generic key management services.
+This document describes conceptually the relationship between application and key management services.
 
-Note that this module is not related to "storing digital objects" that are cryptographic keys.
+## KMS Principles
 
-It is referring to the internal interfaces needed to securely manage "digital object".
+- **Isolatation**: The application and the key management service are isolated from each other. The application should not be able to access the key management service directly or shared runtime resources. The application run-time space should be considered **UNTRUSTED**.
+    
+    - Isolation Strategies:
 
+        - **In-memory** Run-time:
+            - `Trusted Execution Environment` - Dedicated run-time space where it only runs trusted pre-compiled vetted code. They should have tampering resistance and detection mechanisms. (i.e [Trusty TEE](https://source.android.com/security/trusty))
+            - `Processes` - The application and the key management service run in separate processes.
+            - `Containers & Virtualization` - The application and the key management service run in separate containers. (i.e [HashiCorp Vault](https://www.vaultproject.io/))
+        - **Hardware**:
+            - `HSM` - The application and the key management service run on separate hardware. (i.e [HSM](https://en.wikipedia.org/wiki/Hardware_security_module))
+            - `Ledger & Trezor` - Established hardware wallets in the cryptocurrency space. But integration is vendor specific and **NOT** standardized or security compliance friendly. Also, addition of new fetures is dependent on the vendor. (i.e [Ledger](https://www.ledger.com/), [Trezor](https://trezor.io/))
+            - `Over the Network` - The application and the key management service run on separate cloud HSMs. (i.e [AWS CloudHSM](https://aws.amazon.com/cloudhsm/))
+
+- **Access Control**: The application should only be able to access the key management service through a well defined interface. The key management service should be able to control access to its resources.
+
+- **Tampering Detection**: The key management service should be able to detect if the application has been tampered with.
```

Comment:
Should the KMS be able to determine if the application has been tampered with (as stated) or if the application or some other system has tampered with the KMS?